Use go client sdk in HelloWorld example (#564)
hyperledger-labs/fabric-private-chaincode: examples/hallowelt/go_app/README.md

> # HelloWorld tutorial using Go client Fabric SDK

Reviewer: ```suggestion
# HelloWorld tutorial using FPC Client SDK for Go
```

Let's try to call it our go sdk consistently "FPC Client SDK for Go"

Author: ok

> This tutorial shows how to invoke Fabric Private Chaincode using the Go Fabric Client SDK and the Fabric-Private-chaincode(FPC) framework.  This assumes familiarity with the [concepts](https://hyperledger-fabric.readthedocs.io/en/release-2.0/whatis.html#), [the programming model](https://hyperledger-fabric.readthedocs.io/en/release-2.0/chaincode.html) and the [gateway] (https://hyperledger-fabric.readthedocs.io/en/release-2.0/developapps/gateway.html) in Hyperledger Fabric v2.0.

Reviewer: ```suggestion
This tutorial shows how to interact with a FPC Chaincode using the FPC Client SDK for Go.  This assumes familiarity with the [concepts](https://hyperledger-fabric.readthedocs.io/en/release-2.2/whatis.html#), [the programming model](https://hyperledger-fabric.readthedocs.io/en/release-2.2/chaincode.html) and the [gateway](https://hyperledger-fabric.readthedocs.io/en/release-2.2/developapps/gateway.html) in Hyperledger Fabric v2.2.
```

Reviewer: In this intro we should also link to our FPC Client SDK docu and give a hint that we have more examples in xyz.

Author: ah! good point

> ## Prerequisites

Reviewer: I guess this subsection is not needed when this entire readme got integrated into the hello world tutorial as subsection.

Author: yes probably

Author: ok, remove prerequisites.

## Develop client gateway go script
* Create a folder named `go_app`  in `FPC-INSTALL-DIR/examples/helloworld`.
```bash
cd $GOPATH/src/github.com/hyperledger-labs/fabric-private-chaincode/examples/helloworld
mkdir go_app
cd go_app
touch helloworld.go
```

* Define the go package and import the necessary libraries that will be used. `os` to access operating system variables and two libraries that enable us to access the fabric gateway and call the FPC chaincode from the gateway.

```go
package main

import (
	"os"

	fpc "github.com/hyperledger-labs/fabric-private-chaincode/client_sdk/go/pkg/gateway"
	"github.com/hyperledger-labs/fabric-private-chaincode/integration/client_sdk/go/utils"
	"github.com/hyperledger/fabric-sdk-go/pkg/gateway"
	"github.com/hyperledger/fabric/common/flogging"
)
```

* We will also define several variables. `network` represents the Fabric network and `contract` our fabric private chaincode developed previously. We also are using a variable to be able to capture the errors that the different functions might return.

```go
var (
	network  *gateway.Network
	contract fpc.Contract

> 	err      error

Reviewer: ```suggestion
```

The error is not really needed here.

Author: ok, will define it inside the function and see how it goes. Will do the same with Network and contract.

	logger   = flogging.MustGetLogger("helloworld")
)
```
* We will then add the main function which will execute our code. The environent variable `CC_ID` represents the FPC chaincode already installed on the channel and will be set from the bash script we will create to run evenrything. The Fabric network we are setting up has only one channel called `mychannel` which we will connect to, using the `SetupNetwork` go method provided in the `utils` library we imported before. Then we will access the FPC chaincode and call the `SubmitTransaction` go method that uses the first parameter as the chaincode method to invoke, the second as the name of the asset and the third as the value. In this case we will be storing 100 as the value of `asset1`  invoking the `storeAsset` method.

```go
func main() {
	ccID := os.Getenv("CC_ID")
	logger.Infof("Environment variable CC_ID: %v", ccID)
	// get network

> 	network, err = utils.SetupNetwork("mychannel")

Reviewer: let's do somthing with the error ... I think a panic would be simple and ok here.

Reviewer: Also, should we maybe also read the channel name from ENV?

Author: is it stored already in an ENV? or would we have to configure like with CC_ID?

Reviewer: It think it is not - so exporting as we do for CC_ID would work I guess

	// Get FPC Contract
	contract := fpc.GetContract(network, ccID)

	// Invoke FPC Chaincode storeAsset
	logger.Infof("--> Invoke FPC chaincode storeAsset: ")
	result, err := contract.SubmitTransaction("storeAsset", "asset1", "100")
	if err != nil {
		logger.Fatalf("Failed to Submit transaction: %v", err)
	}
	logger.Infof("--> Result: %s", string(result))
}

> )

Reviewer: ```suggestion

```

```
* To be able to query the ledger we have to call the same method changing the first parameter to `retrieveAsset` and since we are querying there is no need for the third parameter.
```go
// Invoke FPC Chaincode retrieveAsset
	logger.Infof("--> Invoke FPC chaincode retrieveAsset: ")
	result, err = contract.SubmitTransaction("retrieveAsset", "asset1")
	if err != nil {
		logger.Fatalf("Failed to Submit transaction: %v", err)
	}
	logger.Infof("--> Result: %s", string(result))
```
* Both previous transactions have been registered on the ledger, to be able to query without this occurring we will use the `EvaluateTransaction` go method instead.
```go
	// Evaluate FPC Chaincode retrieveAsset
	logger.Infof("--> Evaluate FPC chaincode retrieveAsset: ")
	result, err = contract.EvaluateTransaction("retrieveAsset", "asset1")
	if err != nil {
		logger.Fatalf("Failed to Evaluate transaction: %v", err)
	}
	logger.Infof("--> Result: %s", string(result))
```

> The whole text of the program would be

Reviewer: While reading I had the feeling that we could provide the entire program code in a single snippet rather than splitting it as we did in the hello world tutorial.

```go
package main

import (
	"os"

	fpc "github.com/hyperledger-labs/fabric-private-chaincode/client_sdk/go/pkg/gateway"
	"github.com/hyperledger-labs/fabric-private-chaincode/integration/client_sdk/go/utils"
	"github.com/hyperledger/fabric-sdk-go/pkg/gateway"
	"github.com/hyperledger/fabric/common/flogging"
)

var (
	network  *gateway.Network
	contract fpc.Contract
	err      error
	logger   = flogging.MustGetLogger("helloworld")
)

func main() {
	ccID := os.Getenv("CC_ID")
	logger.Infof("Environment variable CC_ID: %v", ccID)
	// get network
	network, err = utils.SetupNetwork("mychannel")

	// Get FPC Contract
	contract := fpc.GetContract(network, ccID)

	// Invoke FPC Chaincode storeAsset
	logger.Infof("--> Invoke FPC chaincode storeAsset: ")
	result, err := contract.SubmitTransaction("storeAsset", "asset1", "100")
	if err != nil {
		logger.Fatalf("Failed to Submit transaction: %v", err)
	}
	logger.Infof("--> Result: %s", string(result))
	

> 	// Invoke FPC Chaincode retrieveAsset
> 	logger.Infof("--> Invoke FPC chaincode retrieveAsset: ")
> 	result, err = contract.SubmitTransaction("retrieveAsset", "asset1")
> 	if err != nil {
> 		logger.Fatalf("Failed to Submit transaction: %v", err)
> 	}
> 	logger.Infof("--> Result: %s", string(result))
> 	

Reviewer: I would remove this, since retrieveAsset is then also called using EvaluateTransaction
```suggestion

```

Author: but submit is not the same as Evaluate. And also, we are just executing the same transactions that are done in the HelloWorld so that they can compare one and the other

Reviewer: I know, nevertheless we demonstrate SubmitTransaction already with the `storeAsset` function. I think this is good enough - and there is no need to call `retrieveAsset` using Submit and Evaluate. (Maybe the original HelloWorld example is also not good). However, we can keep it for a moment see what Bruno things. :)

	// Evaluate FPC Chaincode retrieveAsset
	logger.Infof("--> Evaluate FPC chaincode retrieveAsset: ")
	result, err = contract.EvaluateTransaction("retrieveAsset", "asset1")
	if err != nil {
		logger.Fatalf("Failed to Evaluate transaction: %v", err)
	}
	logger.Infof("--> Result: %s", string(result))
}
```
* Now we have to test our code. Copy the test.sh bash script from helloworld to this directory. We are going to make some changes to this script. The first changes that need to be made are the reference to the current directory. We have to add one more level of directory `/..` to the variables `FPC_PATH` and `FABRIC_CFG_PATH`. 
```bash
FPC_PATH="${SCRIPTDIR}/../../.."
FABRIC_CFG_PATH="${SCRIPTDIR}/../../../integration/config"
```

> We also need to change the code that invoked the smart contract

Reviewer: While following these instructions, I thought it would be easier to modify the existing `examples/helloworld/test.sh`.

Either we just add the go app in addition to the exising cli calls, or we let the reader remove the cli calls and just add the go app. But having the cli calls and go app would also not harm I believe. WDYT?

Author: I think for someone new to this. having both examples separate can help understand better the similarities and the differences. If we did it in the same one, they would probably have 2 bash test scripts. The unchanged one and the go one ....

Reviewer: mh - let's postpone this for a moment and focus on fixing the rest of the document. Then we come back to this and get some feedback from others. Jordan, Jeb, Mic, .... and then we just do a majority voting :)

    say "- install helloworld chaincode"
    PKG=/tmp/${CC_ID}.tar.gz
    try ${PEER_CMD} lifecycle chaincode package --lang fpc-c --label ${CC_ID} --path ${CC_PATH} ${PKG}
    try ${PEER_CMD} lifecycle chaincode install ${PKG}

    PKG_ID=$(${PEER_CMD} lifecycle chaincode queryinstalled | awk "/Package ID: ${CC_ID}/{print}" | sed -n 's/^Package ID: //; s/, Label:.*$//;p')

    try ${PEER_CMD} lifecycle chaincode approveformyorg -o ${ORDERER_ADDR} -C ${CHAN_ID} --package-id ${PKG_ID} --name ${CC_ID} --version ${CC_VER} --sequence ${CC_SEQ} --signature-policy ${CC_EP}
    try ${PEER_CMD} lifecycle chaincode checkcommitreadiness -C ${CHAN_ID} --name ${CC_ID} --version ${CC_VER} --sequence ${CC_SEQ} --signature-policy ${CC_EP}
    try ${PEER_CMD} lifecycle chaincode commit -o ${ORDERER_ADDR} -C ${CHAN_ID} --name ${CC_ID} --version ${CC_VER} --sequence ${CC_SEQ} --signature-policy ${CC_EP}

    # create an FPC Chaincode enclave
    try ${PEER_CMD} lifecycle chaincode initEnclave -o ${ORDERER_ADDR} --peerAddresses "localhost:7051" --name ${CC_ID}

    say "- invoking FPC chaincode using go client SDK"
    export CC_ID
    GOAPP="./go_app"
    go run $GOAPP

> #    go run .

Reviewer: this is not match the description above. :/

Author: oh! forgot to change there! thanks!

say "Run helloworld  test"

say "- setup ledger"
ledger_init

say "- helloworld test"
helloworld_test

say "- shutdown ledger"
ledger_shutdown

para
yell "Helloworld test PASSED"

exit 0
```
* When you execute from the `go_app` directory:
```bash
	bash test.sh
```
you should get the following output:
```bash
test.sh: - invoking FPC chaincode using go client SDK
2021-04-07 15:31:52.554 UTC [helloworld] main -> INFO 001 Environment variable CC_ID: helloworld_test
 [fabsdk/core] 2021/04/07 15:31:52 UTC - cryptosuite.GetDefault -> INFO No default cryptosuite found, using default SW implementation
2021-04-07 15:31:52.666 UTC [helloworld] main -> INFO 002 --> Invoke FPC chaincode storeAsset: 
2021-04-07 15:31:54.881 UTC [helloworld] main -> INFO 003 --> Result: OK
2021-04-07 15:31:54.881 UTC [helloworld] main -> INFO 004 --> Invoke FPC chaincode retrieveAsset: 
2021-04-07 15:31:56.970 UTC [helloworld] main -> INFO 005 --> Result: asset1:100
2021-04-07 15:31:56.970 UTC [helloworld] main -> INFO 006 --> Evaluate FPC chaincode retrieveAsset: 
2021-04-07 15:31:57.005 UTC [helloworld] main -> INFO 007 --> Result: asset1:100
test.sh: - shutdown ledger

test.sh: Helloworld test PASSED
```

> Congratulations! You have created a go client using the SDK to invoke your FPC chaincode!

Reviewer: ```suggestion
Congratulations! You have created a client app using the FPC Client SDK to invoke your FPC Chaincode!
```
